extensions/librdkafka/KafkaConnection.h

```diff
@@ -0,0 +1,197 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef NIFI_MINIFI_CPP_KAFKACONNECTION_H
+#define NIFI_MINIFI_CPP_KAFKACONNECTION_H
+
+#include <atomic>
+#include <mutex>
+#include <string>
+#include "core/logging/LoggerConfiguration.h"
+#include "core/logging/Logger.h"
+#include "rdkafka.h"
+#include "KafkaTopic.h"
+
+class KafkaConnectionKey {
+public:
+    std::string client_id_;
+    std::string brokers_;
+};
+
+class KafkaConnection {
+ public:
+
+  explicit KafkaConnection(const KafkaConnectionKey &key)
+      : logger_(logging::LoggerFactory<KafkaConnection>::getLogger()),
+        conf_(nullptr),
+        kafka_connection_(nullptr) {
+    // Obtain lease before unlocking KafkaPool.
+    lease_ = true;
+    initialized_ = false;
+    key_ = key;
+  }
+
+  ~KafkaConnection() {
+    remove();
+  }
+
+  void remove() {
+    topics_.clear();
+    removeConnection();
+  }
+
+  void removeConnection() {
+    if (kafka_connection_) {
+      rd_kafka_flush(kafka_connection_, 10 * 1000); /* wait for max 10 seconds */
+      rd_kafka_destroy(kafka_connection_);
+      modifyLoggers([&](std::unordered_map<const rd_kafka_t*, std::weak_ptr<logging::Logger>>& loggers) {
+        loggers.erase(kafka_connection_);
+      });
+      kafka_connection_ = nullptr;
+    }
+    if (conf_) {
+      rd_kafka_conf_destroy(conf_);
+      conf_ = nullptr;
+    }
+    initialized_ = false;
+  }
+
+  bool initialized() {
+    return initialized_;
+  }
+
+  void setConnection(rd_kafka_t *producer, rd_kafka_conf_t *conf) {
+    std::lock_guard<std::mutex> lock(mutex_);
+    removeConnection();
+    kafka_connection_ = producer;
+    conf_ = conf;
+    initialized_ = true;
+    modifyLoggers([&](std::unordered_map<const rd_kafka_t*, std::weak_ptr<logging::Logger>>& loggers) {
+      loggers[producer] = logger_;
+    });
+  }
+
+  rd_kafka_conf_t *getConf() {
+    return conf_;
+  }
+
+  rd_kafka_t *getConnection() {
+    return kafka_connection_;
+  }
+
+  bool hasTopic(const std::string &topic) {
+    return topics_.find(topic) != std::end(topics_);
+  }
+
+  std::shared_ptr<KafkaTopic> getTopic(const std::string &topic) {
+    auto topicObj = topics_.find(topic);
+    if (topicObj != topics_.end()) {
+      return topicObj->second;
+    }
+    return nullptr;
+  }
+
+  KafkaConnectionKey const * const getKey() const {
+    return &key_;
+  }
+
+  void putTopic(const std::string &topicName, const std::shared_ptr<KafkaTopic> topic) {
+    topics_.insert(std::make_pair(topicName, topic));
```

Comment:
Why not simply topics_[topicName] = topic ?
On the other side, your the const shared ptr hardly makes sense, either const& (which doesn't copy, so the counter are not incremented), or simply non-const instance.